TombEngine/Objects/TR3/Trap/ElectricCleaner.cpp

```diff
@@ -0,0 +1,476 @@
+						r->floor[((z - r->z) / BLOCK(1)) + r->xSize * ((x - r->x) / BLOCK(1))].Stopper = 1;
+					}
+
+					break;
+
+				// Facing X-
+				case ANGLE(-90.0f):
+					x = item.Pose.Position.x;
+					z = item.Pose.Position.z - BLOCK(1);
+					CheckCleanerHeading(item, x, y, z, item.RoomNumber, left);
+
+					x = item.Pose.Position.x - BLOCK(1);
+					z = item.Pose.Position.z;
+					CheckCleanerHeading(item, x, y, z, item.RoomNumber, ahead);
+
+					if (!ahead && !left && item.ItemFlags[0] > 0)
+					{
+						item.Pose.Orientation.y += ELECTRIC_CLEANER_TURN_RATE;
+						item.ItemFlags[0] = ELECTRIC_CLEANER_TURN_RATE;
+					}
+					else if (!ahead && !left && item.ItemFlags[0] < 0)
+					{
+						item.Pose.Orientation.y -= ELECTRIC_CLEANER_TURN_RATE;
+						item.ItemFlags[0] -= ELECTRIC_CLEANER_TURN_RATE;
+					}
+					else if (left && item.ItemFlags[0] > 0)
+					{
+						item.Pose.Orientation.y -= ELECTRIC_CLEANER_TURN_RATE;
+						item.ItemFlags[0] = -ELECTRIC_CLEANER_TURN_RATE;
+					}
+					else
+					{
+						item.ItemFlags[0] = ELECTRIC_CLEANER_TURN_RATE;
+						item.ItemFlags[1] = 1;
+						item.Pose.Position.x -= item.ItemFlags[2];
+						x = item.Pose.Position.x - BLOCK(1);
+						y = item.Pose.Position.y;
+						z = item.Pose.Position.z;
+						roomNumber = item.RoomNumber;
+						floor = GetFloor(x, y, z, &roomNumber);
+						r = &g_Level.Rooms[roomNumber];
+						r->floor[((z - r->z) / BLOCK(1)) + r->xSize * ((x - r->x) / BLOCK(1))].Stopper = 1;
+					}
+
+					break;
+
+				// Facing Z-
+				case ANGLE(-180.0f):
+					x = item.Pose.Position.x + BLOCK(1);
+					z = item.Pose.Position.z;
+					CheckCleanerHeading(item, x, y, z, item.RoomNumber, left);
+
+					x = item.Pose.Position.x;
+					z = item.Pose.Position.z - BLOCK(1);
+					CheckCleanerHeading(item, x, y, z, item.RoomNumber, ahead);
+
+					if (!ahead && !left && item.ItemFlags[0] > 0)
+					{
+						item.Pose.Orientation.y += ELECTRIC_CLEANER_TURN_RATE;
+						item.ItemFlags[0] = ELECTRIC_CLEANER_TURN_RATE;
+					}
+					else if (!ahead && !left && item.ItemFlags[0] < 0)
+					{
+						item.Pose.Orientation.y -= ELECTRIC_CLEANER_TURN_RATE;
+						item.ItemFlags[0] -= ELECTRIC_CLEANER_TURN_RATE;
+					}
+					else if (left && item.ItemFlags[0] > 0)
+					{
+						item.Pose.Orientation.y -= ELECTRIC_CLEANER_TURN_RATE;
+						item.ItemFlags[0] = -ELECTRIC_CLEANER_TURN_RATE;
+					}
+					else
+					{
+						item.ItemFlags[0] = ELECTRIC_CLEANER_TURN_RATE;
+						item.ItemFlags[1] = 1;
+						item.Pose.Position.z -= item.ItemFlags[2];
+						x = item.Pose.Position.x;
+						y = item.Pose.Position.y;
+						z = item.Pose.Position.z - BLOCK(1);
+						roomNumber = item.RoomNumber;
+						floor = GetFloor(x, y, z, &roomNumber);
+						r = &g_Level.Rooms[roomNumber];
+						r->floor[((z - r->z) / BLOCK(1)) + r->xSize * ((x - r->x) / BLOCK(1))].Stopper = 1;
+					}
+
+					break;
+				}
+
+				TestTriggers(&item, true);
+			}
+			else
+			{
+				// No new target, keep updating position.
```

Comment:
The cleaner is moving with its own calculated movement in control function, maybe it could benefit of the same locomotion system that other entities use for movement?

Reply:
I don't think it's a good idea. Cleaner is very fragile, it tends to break or travel to incorrect directions with even slight change to movement or rotation code. I'd leave it as is.

Reply:
That's oks too.